cassiopeia/simulator/Cas9LineageTracingDataSimulator.py

```diff
@@ -0,0 +1,202 @@
+            
+            # silence cassettes
+            character_array = self.silence_cassettes(character_array)
+
+            tree.set_character_states(node, character_array)
+
+
+    def collapse_sites(self, character_array: List[int], cuts: List[int]) -> Tuple[List[int], List[int]]:
+        """Collapses cassettes.
+
+        Given a character array and a new set of cuts that Cas9 is inducing,
+        this function will infer which cuts occur within a given cassette and
+        collapse the sites between the two cuts.
+
+        Args:
+            character_array: Character array in progress
+            cuts: Sites in the character array that are being cut.
+
+        Returns:
+            The updated character array and sites that are not part of a
+                cassette collapse.
+        """
+
+        cassettes = self.get_cassettes()
+        cut_to_cassette = np.digitize(cuts, cassettes)
+        cuts_remaining =[]
+        for cassette in np.unique(cut_to_cassette):
+            cut_indices = np.where(cut_to_cassette == cassette)[0]
+            if len(cut_indices) > 1:
+                cuts = new_cuts[cut_indices]
+                left, right = np.min(cuts), np.max(cuts)
+                for site in range(left, right+1):
+                    character_array[site] = -1
+            else:
+                cuts_remaining.append(cut_indices[0])
+        
+        return character_array, cuts_remaining
+
+    def introduce_states(self, character_array: List[int], cuts: List[int]) -> List[int]:
+        """Adds states to character array.
+
+        New states are added to the character array at the predefined cut
+        locations.
+
+        Args:
+            character_array: Character array
+            cuts: Loci being cut
+
+        Returns:
+            An updated character array.
+        """
+
+        for i in cuts:
+            state = self.random.choice(list(self.mutation_priors.keys()), 1, p=list(self.mutation_priors.values()))
+            character_array[i] = state
+
+        return character_array
+
+    def silence_cassettes(self, character_array: List[int]):
+        """Silences cassettes.
+
+        Using the predefined silencing rate of this simulator, this function
+        simulates the rare event of transcriptional silencing.
+
+        Args:
+            character_array: Character array
+
+        Returns:
+            An updated character array.
+
+        """
+        cassettes = self.get_cassettes()
+        cut_site_by_cassette = np.digitize(range(len(character_array)), cassettes)
+        for cassette in cassettes:
+            if np.random.uniform() < self.silencing_rate:
+                indices = np.where(cut_site_by_cassette)
+                left, right = np.min(indices), np.max(indices)
+                for site in range(left, right+1):
+                    character_array[site] = -1
+
+    def get_cassettes(self) -> List[int]:
```

Comment:
Should this (and other methods) be made private?

Reply:
I prefer to keep these public for testing purposes. Is there a strong reason for why we'd not like a user to have access to these methods?

Reply:
No strong reason, other than when reading code I like being able to distinguish key methods (public) from implementation details (private). But I can also infer this by looking at the abstract base class we are implementing, so not really an issue. In this case, you are right that there might be a reason for a user to want to use this method.

Reply:
Okay, sounds good! Let's keep them public for now.